### DISCUSSION TITLE
Polygon simplification with penalty for sharp corners

### DISCUSSION THREAD

**Discussion by wipfli** [onthegomap/planetiler#708] [Q&A]
When doing polygon simplification, I often get sharp corners when I increase the pixel tolerance. I personally find this looks bad.

Is there a method which gives a penalty for sharp corners when doing the polygon simplification?

I would be happy to trade off area accuracy for fewer sharp corners...

**Comment by wipfli** [comment]
Wang-Müller might be a good choice. Playing around at the moment with a QGIS plugin that implements it: https://github.com/NRCan/geo_sim_processing

**Reply by msbarry** [reply]
There's also [visvalingham wyatt](https://en.wikipedia.org/wiki/Visvalingam–Whyatt_algorithm) meant to smooth out edges. I believe natural earth uses visvalingham wyatt with effective area from https://mapshaper.org to generalize shapes at low zooms

**Comment by wipfli** [comment]
Love esri docs...

![Screenshot_20231104-190016](https://github.com/onthegomap/planetiler/assets/53421382/8b52e635-3b1e-411b-a156-092813dbe598)

**Reply by msbarry** [reply]
Nice, those are good names that describe what they do too. In general I think planetiler should have flexible support for multiple kinds of simplification/smoothing with more flexible options, or maybe even pipelines that compose multiple operations like this in a row.

**Reply by wipfli** [reply]
Oh yeah a geometry pipeline could be fun!

**Comment by wipfli** [comment]
For polygonized raster data an easy way could also be to do a bit of midpoint smoothing, see https://github.com/wipfli/midpoint-smoothing :

![image](https://github.com/onthegomap/planetiler/assets/53421382/6dfcc815-d0fd-4dd5-ae20-af0fcc4cd388)

**Comment by brawer** [comment]
IGN, the French National Institute of Geographic and Forest Information, implemented various cartographic algorithms in Java. See their [web page](https://ignf.github.io/CartAGen/docs/algorithms.html) and [code repository](https://github.com/IGNF/CartAGen/tree/master/cartagen-core/src/main/java/fr/ign/cogit/cartagen/algorithms). They appear to have switched to Python, see [here](https://cartagen.readthedocs.io/en/latest/index.html), but the Java code might still be interesting to read.
